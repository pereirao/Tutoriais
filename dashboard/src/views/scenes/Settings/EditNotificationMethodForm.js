// @flow

import React, {PureComponent} from 'react';
import {connect} from 'react-redux';
import style_sheet from './Settings.module.css';
import MBTable from '../../elements/MBTable';
import MBButton from '../../elements/MBButton';
import MBIcon from '../../elements/MBIcon';
import {notification_method_actions, notification_method_helpers} from '../../../business/notification_method';
import I18n from '../../../localization';


const {NOTIFICATION_METHOD_COLUMN_ORDER} = notification_method_helpers;

export class EditNotificationMethodForm extends PureComponent {
  constructor(props: Object) {
    super(props);
    const {notification_method} = this.props;
    this.state = {
      id: notification_method.id,
      notification_type: notification_method.notification_type,
      value: notification_method.value,
      label: notification_method.label || '',
      active: notification_method.active
    };
  }

  deleteNotificationMethod = () => {
    if (window.confirm(I18n.t('ui.alert.confirm_delete_notification_method'))) {
      this.props.destroyNotificationMethod(this.state.id);
      this.props.toggleEditMode();
    }
  }

  onChange = (e, key) => {
    this.setState({
      [key]: e.target.value
    });
  }

  submitNotificationMethod = () => {
    this.props.toggleEditMode();
    this.props.updateNotificationMethod({...this.state});
  };

  notificationTypeInput = (key, notification_method) => (
    <select
      onChange={(e) => this.onChange(e, key)}
      className={`notification_method_${notification_method.id}`}
      id={`${notification_method.id}${key}`}
      value={this.state[key]}>
      <option value="phone">Phone</option>
      <option value="sms">SMS</option>
      <option value="fax">Fax</option>
      <option value="email">Email</option>
    </select>)

  notificationValueInput = (key, notification_method) => {
    return (
      <input
        onChange={(e) => this.onChange(e, key)}
        className={`notification_method_${notification_method.id}`}
        id={`${notification_method.id}${key}`}
        value={this.state[key]}
        type="text"/>
    );
  }

  notificationActiveInput = (key, notification_method) => (
    <select
      onChange={(e) => this.onChange(e, key)}
      className={`notification_method_${notification_method.id}`}
      id={`${notification_method.id}${key}`}
      value={this.state[key]}>
      <option value>Active</option>
      <option value={false}>Inactive</option>
    </select>
  )

  notificationEditButton = (isMobile) => (
    <div className={style_sheet.editOptionsWrapper}>
      <div className={style_sheet.actionsWrapper}>
        <MBButton button_type="link" onClick={() => {
          this.submitNotificationMethod();
        }}>
          {I18n.t('ui.table.submit')}
        </MBButton>
        <div className={style_sheet.deleteIcon}>
          <MBButton button_type="link" onClick={() => {
            this.deleteNotificationMethod();
          }}>
            <MBIcon icon="delete" color="mb_red"/>
          </MBButton>
        </div>
      </div>
      <div className={isMobile ? style_sheet.mobileCancelWrapper : style_sheet.cancelWrapper}>
        <MBIcon icon="cancel" color="mb_red" onClick={() => {
          this.props.toggleEditMode();
        }}/>
      </div>
    </div>
  )

  // Value should be more specific to notification type and have validations
  formatCellDatum = (key, notification_method) => {
    switch (key) {
      case 'notification_type':
        return this.notificationTypeInput(key, notification_method);
      case 'value':
      case 'label':
        return this.notificationValueInput(key, notification_method);
      case 'active':
        return this.notificationActiveInput(key, notification_method);
      case 'edit':
        return this.notificationEditButton();
      case 'mobile_summary': {
        return (
          <table className={style_sheet.notificationTable}>
            <tbody>
            <tr>
              <td>Type:</td>
              <td>
                {this.notificationTypeInput('notification_type', notification_method)}
              </td>
            </tr>
            <tr>
              <td>Label:</td>
              <td>{this.notificationValueInput('label', notification_method)}</td>
            </tr>
            <tr>
              <td>Value:</td>
              <td>{this.notificationValueInput('value', notification_method)}</td>
            </tr>
            <tr>
              <td>Status:</td>
              <td>{this.notificationActiveInput('active', notification_method)}</td>
            </tr>
            <tr>
              <td colSpan={2}>{this.notificationEditButton(true)}</td>
            </tr>
            </tbody>
          </table>
        );
      }
      default:
        return <p>{this.state[key]}</p>;
    }
  };

  render() {
    const {notification_method} = this.props;
    return (
      <MBTable.Row key={`${notification_method.id}row`}>
        {NOTIFICATION_METHOD_COLUMN_ORDER.map(col => {
          return (
            <MBTable.Cell
              key={`${notification_method.id}${col}`}
              mobile_only={col === 'mobile_summary'}
              mobile_hidden={col !== 'mobile_summary'}>
              {this.formatCellDatum(col, notification_method)}
            </MBTable.Cell>
          );
        })}
      </MBTable.Row>
    );
  }
}

const EditNotificationMethodFormDTP = {
  updateNotificationMethod: notification_method_actions.updateNotificationMethod,
  destroyNotificationMethod: notification_method_actions.destroyNotificationMethod
};
const EditNotificationMethodFormContainer = connect(null, EditNotificationMethodFormDTP)(EditNotificationMethodForm);

export default EditNotificationMethodFormContainer;
