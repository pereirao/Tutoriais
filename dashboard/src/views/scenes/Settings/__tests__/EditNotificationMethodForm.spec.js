import React from 'react';
import { EditNotificationMethodForm } from '../EditNotificationMethodForm';
import notification_method_factory from '../../../../business/notification_method/__tests__/notification_method.factory';


it('renders correctly', () => {
  expect(shallow(<EditNotificationMethodForm notification_method={notification_method_factory.build()} />)).toMatchSnapshot();
});
