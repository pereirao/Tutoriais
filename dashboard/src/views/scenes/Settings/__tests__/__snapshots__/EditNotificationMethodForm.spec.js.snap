// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Row>
  <Cell
    mobile_hidden={false}
    mobile_only={true}
  >
    <p />
  </Cell>
  <Cell
    mobile_hidden={true}
    mobile_only={false}
  >
    <select
      className="notification_method_1"
      id="1notification_type"
      onChange={[Function]}
    >
      <option
        value="phone"
      >
        Phone
      </option>
      <option
        value="sms"
      >
        SMS
      </option>
      <option
        value="fax"
      >
        Fax
      </option>
      <option
        value="email"
      >
        Email
      </option>
    </select>
  </Cell>
  <Cell
    mobile_hidden={true}
    mobile_only={false}
  >
    <input
      className="notification_method_1"
      id="1value"
      onChange={[Function]}
      type="text"
      value="1234567890"
    />
  </Cell>
  <Cell
    mobile_hidden={true}
    mobile_only={false}
  >
    <input
      className="notification_method_1"
      id="1label"
      onChange={[Function]}
      type="text"
      value="default_factory_notification_method"
    />
  </Cell>
  <Cell
    mobile_hidden={true}
    mobile_only={false}
  >
    <select
      className="notification_method_1"
      id="1active"
      onChange={[Function]}
      value={true}
    >
      <option
        value={true}
      >
        Active
      </option>
      <option
        value={false}
      >
        Inactive
      </option>
    </select>
  </Cell>
  <Cell
    mobile_hidden={true}
    mobile_only={false}
  >
    <div
      className="editOptionsWrapper"
    >
      <div
        className="actionsWrapper"
      >
        <MBButton
          button_type="link"
          onClick={[Function]}
        >
          Submit
        </MBButton>
        <div
          className="deleteIcon"
        >
          <MBButton
            button_type="link"
            onClick={[Function]}
          >
            <MBIcon
              color="mb_red"
              icon="delete"
            />
          </MBButton>
        </div>
      </div>
      <div
        className="cancelWrapper"
      >
        <MBIcon
          color="mb_red"
          icon="cancel"
          onClick={[Function]}
        />
      </div>
    </div>
  </Cell>
</Row>
`;
